File: src/01.py
import sys
from collections import Counter

def parse_input():
    # Read the input from the stdinput
    lines = sys.stdin.readlines()

    # Initialize two lists for left and right numbers
    left_numbers = []
    right_numbers = []

    # Parse the input into the two lists
    for line in lines:
        if line.strip():  # Ensure we don't process empty lines
            left, right = map(int, line.split())
            left_numbers.append(left)
            right_numbers.append(right)

    return left_numbers, right_numbers

def calculate_total_distance(left_numbers, right_numbers):
    # Sort both lists
    left_numbers.sort()
    right_numbers.sort()

    # Calculate the total distance
    total_distance = sum(abs(l - r) for l, r in zip(left_numbers, right_numbers))

    return total_distance

def calculate_similarity_score(left_numbers, right_numbers):
    # Count occurrences of each number in the right list
    right_count = Counter(right_numbers)

    # Calculate the total similarity score
    total_similarity_score = sum(l * right_count[l] for l in left_numbers)

    return total_similarity_score

def main():
    left_numbers, right_numbers = parse_input()

    total_distance = calculate_total_distance(left_numbers, right_numbers)
    print(f"The total distance between the lists is: {total_distance}")

    total_similarity_score = calculate_similarity_score(left_numbers, right_numbers)
    print(f"The total similarity score between the lists is: {total_similarity_score}")

if __name__ == "__main__":
    main()

File: src/01.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_SIZE 100000

typedef struct {
    int key;
    int value;
} HashEntry;

typedef struct {
    HashEntry entries[MAX_SIZE];
    size_t size;
} HashMap;

static inline void hashmap_init(HashMap* map) {
    for (size_t i = 0; i < MAX_SIZE; ++i) {
        map->entries[i].key = -1; // Mark as unused
    }
    map->size = 0;
}

static inline int hash(int key) {
    return abs(key) % MAX_SIZE;
}

static inline void hashmap_insert(HashMap* map, int key, int value) {
    int index = hash(key);
    while (map->entries[index].key != -1 && map->entries[index].key != key) {
        index = (index + 1) % MAX_SIZE; // Linear probing
    }
    map->entries[index].key = key;
    map->entries[index].value = value;
}

static inline int hashmap_find(HashMap* map, int key) {
    int index = hash(key);
    while (map->entries[index].key != -1 && map->entries[index].key != key) {
        index = (index + 1) % MAX_SIZE; // Linear probing
    }
    return (map->entries[index].key == key) ? map->entries[index].value : 0;
}

static inline void vector_init(int* arr, size_t* size) {
    *size = 0;
}

static inline void vector_push_back(int* arr, int value, size_t* size) {
    arr[(*size)++] = value;
}

int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

int main() {
    int left, right;
    int leftList[MAX_SIZE], rightList[MAX_SIZE];
    size_t leftSize = 0, rightSize = 0;
    HashMap frequencyMap;

    hashmap_init(&frequencyMap);

    // Read each pair of numbers directly into variables
    while (scanf("%d %d", &left, &right) == 2) {
        vector_push_back(leftList, left, &leftSize);
        vector_push_back(rightList, right, &rightSize);
        hashmap_insert(&frequencyMap, right, hashmap_find(&frequencyMap, right) + 1);
    }

    // Sort both lists for total distance calculation
    qsort(leftList, leftSize, sizeof(int), compare);
    qsort(rightList, rightSize, sizeof(int), compare);

    // Calculate the total distance
    int totalDistance = 0;
    for (size_t i = 0; i < leftSize && i < rightSize; ++i) {
        totalDistance += abs(leftList[i] - rightList[i]);
    }

    // Calculate the similarity score
    long long similarityScore = 0;
    for (size_t i = 0; i < leftSize; ++i) {
        int num = leftList[i];
        int count = hashmap_find(&frequencyMap, num);
        if (count > 0) {
            similarityScore += (long long)num * count;
        }
    }

    // Output the results
    printf("Total Distance: %d\n", totalDistance);
    printf("Similarity Score: %lld\n", similarityScore);

    return 0;
}